p2p/dhtdiscovery/discovery.go

```diff
@@ -270,3 +322,175 @@ func (d *Discovery) newDht(ctx context.Context, h host.Host, public, server bool
+		}
+		if connected := len(d.h.Network().Peers()); connected >= d.minPeers {
+			d.backup = nil // once got enough peers no need to keep backup, they are either already connected or unavailable
+			d.logger.Debug("node is connected with required number of peers. skipping bootstrap",
+				zap.Int("required", d.minPeers),
+				zap.Int("connected", connected),
+			)
+		} else {
+			d.connect(ctx, &connEg, d.backup)
+			// no reason to spend more resources if we got enough from backup
+			if connected := len(d.h.Network().Peers()); connected >= d.minPeers {
+				continue
+			}
+			d.connect(ctx, &connEg, d.bootnodes)
+			d.bootstrap(ctx)
+		}
+	}
+}
+
+func (d *Discovery) peerHasTag(p peer.ID) bool {
+	ti := d.h.ConnManager().GetTagInfo(p)
+	if ti == nil {
+		return false
+	}
+	_, found := ti.Tags[discoveryTag]
+	return found
+}
+
+func (d *Discovery) advertiseNS(ctx context.Context, ns string, active func() bool) error {
+	for {
+		var ttl time.Duration
+		if active == nil || active() {
+			var err error
+			d.logger.Debug("advertising for routing discovery", zap.String("ns", ns))
+			ttl, err = d.disc.Advertise(ctx, ns, p2pdisc.TTL(d.advertiseInterval))
+			if err != nil {
+				d.logger.Error("failed to re-advertise for discovery", zap.String("ns", ns), zap.Error(err))
+				ttl = advertiseRetryInterval
+			}
+		} else {
+			ttl = d.advertiseInterval
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(ttl):
+		}
+	}
+}
+
+func (d *Discovery) discoverPeers(ctx context.Context) error {
+	for p := range d.findPeersContinuously(ctx, discoveryNS) {
+		wasSuspended := false
+		for !d.h.NeedPeerDiscovery() {
+			wasSuspended = true
+			d.logger.With().Info("suspending routing discovery",
+				zap.Duration("delay", discoveryHighPeersDelay))
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-time.After(discoveryHighPeersDelay):
+			}
+		}
+		if wasSuspended {
+			d.logger.With().Info("resuming routing discovery")
+		}
+
+		if p.ID == d.h.ID() {
+			continue
+		}
+		freshlyConnected := !d.peerHasTag(p.ID)
+		// trim open conns to free up some space
+		d.h.ConnManager().TrimOpenConns(ctx)
+		// tag peer to prioritize it over the peers found by other means
+		d.h.ConnManager().TagPeer(p.ID, discoveryTag, discoveryTagValue)
+		if d.h.Network().Connectedness(p.ID) != network.Connected {
+			if _, err := d.h.Network().DialPeer(ctx, p.ID); err != nil {
+				d.logger.Error("error dialing peer", zap.Any("peer", p),
+					zap.Error(err))
+				continue
+			}
+			freshlyConnected = true
+		}
+		if freshlyConnected {
+			d.logger.Info("found peer via rendezvous", zap.Any("peer", p))
+		}
+	}
+
+	return nil
+}
+
+func (d *Discovery) discoverRelays(ctx context.Context) error {
+	for p := range d.findPeersContinuously(ctx, relayNS) {
+		if len(p.Addrs) != 0 {
+			d.logger.Debug("found relay candidate", zap.Any("p", p))
+			select {
+			case d.relayCh <- p:
+			case <-ctx.Done():
+				return nil
+			}
+		}
+	}
+	return nil
+}
```

Comment:
It never returns a non-nil error. How about changing to:
```suggestion
func (d *Discovery) discoverRelays(ctx context.Context) {
	for p := range d.findPeersContinuously(ctx, relayNS) {
		if len(p.Addrs) != 0 {
			d.logger.Debug("found relay candidate", zap.Any("p", p))
			select {
			case d.relayCh <- p:
			case <-ctx.Done():
				return
			}
		}
	}
}
```